ddtrace/profiling/collector/_memalloc.c

```diff
@@ -42,113 +42,81 @@ static PyObject* object_string = NULL;
-   functions!
- */
-static memlock_t g_memalloc_lock;
-
 static alloc_tracker_t* global_alloc_tracker;
 
-// This is a multiplatform way to define an operation to happen at static initialization time
-static void
-memalloc_init(void);
-
-static void
-memalloc_prefork(void)
-{
-    // Lock the mutex prior to forking. This ensures that the memory profiler
-    // data structures will be in a consistent state in the child process.
-    // The rest of the memalloc calls do trylock so we don't run the risk
-    // of deadlocking if some other fork handler allocates
-    memlock_lock(&g_memalloc_lock);
-}
-
-static void
-memalloc_postfork_parent(void)
-{
-    memlock_unlock(&g_memalloc_lock);
-}
-
-static void
-memalloc_postfork_child(void)
-{
-    memlock_unlock(&g_memalloc_lock);
-}
-
-#ifdef _MSC_VER
-#pragma section(".CRT$XCU", read)
-__declspec(allocate(".CRT$XCU")) void (*memalloc_init_func)(void) = memalloc_init;
-
-#elif defined(__GNUC__) || defined(__clang__)
-__attribute__((constructor))
-#else
-#error Unsupported compiler
-#endif
-static void
-memalloc_init()
-{
-    memlock_init(&g_memalloc_lock);
-#ifndef _WIN32
-    pthread_atfork(memalloc_prefork, memalloc_postfork_parent, memalloc_postfork_child);
-#endif
-}
-
 static void
 memalloc_add_event(memalloc_context_t* ctx, void* ptr, size_t size)
 {
-    uint64_t alloc_count = atomic_add_clamped(&global_alloc_tracker->alloc_count, 1, ALLOC_TRACKER_MAX_COUNT);
+    /* Safety check: is profiling still enabled? */
+    if (!global_alloc_tracker)
+        return;
 
     /* Return if we've reached the maximum number of allocations */
-    if (alloc_count == 0)
+    if (global_alloc_tracker->alloc_count == ALLOC_TRACKER_MAX_COUNT)
         return;
 
-    // Return if we can't take the guard
-    if (!memalloc_take_guard()) {
-        return;
-    }
+    /* Incrementing the allocation count before taking the guard gives
+     * an accurate count, but means we'll overweight the allocations we
+     * do sample relative to their actual size since our count necessarily
+     * includes allocations we won't sample. We could consider moving this
+     * below the guard and instead adding a placeholder "self allocation"
+     * entry to represent the allocations we intentionally don't sample.
+     */
+    uint64_t alloc_count = global_alloc_tracker->alloc_count++;
 
-    // In this implementation, the `global_alloc_tracker` isn't intrinsically protected.  Before we read or modify,
-    // take the lock.  The count of allocations is already forward-attributed elsewhere, so if we can't take the lock
-    // there's nothing to do.
-    if (!memlock_trylock(&g_memalloc_lock)) {
+    /* Return if we can't take the guard, since we're not going to be able to sample */
+    if (!memalloc_take_guard()) {
         return;
     }
 
     /* Determine if we can capture or if we need to sample */
-    if (global_alloc_tracker->allocs.count < ctx->max_events) {
+    bool should_sample = false;
```

Comment:
Don't think you should actually do this, but I couldn't resist redoing this into a one-liner
```C
bool should_sample = alloc_count < ctx->max_events || random_range(alloc_count) < ctx->max_events;
```